packages/shared/src/hooks/useBookmarkPost.ts

```diff
@@ -1,82 +1,220 @@
-import { AnalyticsEvent } from './analytics/useAnalyticsQueue';
+import { useRequestProtocol } from './useRequestProtocol';
+import AuthContext from '../contexts/AuthContext';
+import { updatePostCache } from './usePostById';
+import { AuthTriggers } from '../lib/auth';
+import { AnalyticsEvent, Origin } from '../lib/analytics';
+import {
+  PostAnalyticsEventFnOptions,
+  optimisticPostUpdateInFeed,
+  postAnalyticsEvent,
+} from '../lib/feed';
+import { FeedItem, PostItem, UpdateFeedPost } from './useFeed';
 
-type UseBookmarkPostParams<T> = {
-  onBookmarkMutate: MutateFunc<T>;
-  onRemoveBookmarkMutate: MutateFunc<T>;
-  onBookmarkTrackObject?: () => AnalyticsEvent;
-  onRemoveBookmarkTrackObject?: () => AnalyticsEvent;
+export type ToggleBookmarkProps = {
+  origin: Origin;
+  post: Post | ReadHistoryPost;
+  opts?: PostAnalyticsEventFnOptions;
 };
-type UseBookmarkPostRet<T> = {
-  bookmark: (variables: T) => Promise<void>;
-  bookmarkToast: (targetState: boolean) => void;
-  removeBookmark: (variables: T) => Promise<void>;
+
+export const bookmarkMutationKey = ['post', 'mutation'];
+
+export type UseBookmarkPostMutationProps = {
+  id?: string;
+  mutation?: string;
+  payload?: Record<string, unknown>;
 };
 
-export default function useBookmarkPost<
-  T extends { id: string } = { id: string },
->({
-  onBookmarkMutate,
-  onRemoveBookmarkMutate,
-  onBookmarkTrackObject,
-  onRemoveBookmarkTrackObject,
-}: UseBookmarkPostParams<T>): UseBookmarkPostRet<T> {
-  const { trackEvent } = useContext(AnalyticsContext);
+export type UseBookmarkPostRollback = () => void;
+
+export type BookmarkProps = {
+  id: string;
+};
+
+export type UseBookmarkPostProps = {
+  onMutate?: (
+    props: UseBookmarkPostMutationProps,
+  ) => UseBookmarkPostRollback | undefined;
+  variables?: unknown;
+};
+
+const prepareVotePostAnalyticsOptions = ({
+  origin,
+  opts,
+}: ToggleBookmarkProps): PostAnalyticsEventFnOptions => {
+  const { extra, ...restOpts } = opts || {};
+
+  const analyticsOptions: PostAnalyticsEventFnOptions = {
+    ...restOpts,
+    extra: { ...extra, origin },
+  };
+
+  return analyticsOptions;
+};
+
+export type UseBookmarkPost = {
+  toggleBookmark: (props: ToggleBookmarkProps) => Promise<void>;
+};
+
+const useBookmarkPost = ({
+  onMutate,
+  variables,
+}: UseBookmarkPostProps = {}): UseBookmarkPost => {
+  const { requestMethod } = useRequestProtocol();
+  const client = useQueryClient();
   const { displayToast } = useToastNotification();
-  const { mutateAsync: bookmark } = useMutation<
-    void,
-    unknown,
-    T,
-    (() => void) | undefined
-  >(
-    ({ id }) =>
-      request(graphqlUrl, ADD_BOOKMARKS_MUTATION, {
-        data: { postIds: [id] },
+  const { user, showLogin } = useContext(AuthContext);
+  const { trackEvent } = useContext(AnalyticsContext);
+
+  const defaultOnMutate = ({ id }) => {
+    updatePostCache(client, id, (post) => ({ bookmarked: !post.bookmarked }));
```

Comment:
Don't you already send the swapped value in some places? (just checking)

Reply:
This is for the default mutation, i.e. when there is no feedQueryKey, so it falls back to the mutation for a single page

Reply:
Ah got it! Thanks for the explanation